migrations/000001_init_schema.up.sql

```diff
@@ -0,0 +1,32 @@
+CREATE TABLE IF NOT EXISTS users (
+    id uuid NOT NULL DEFAULT uuid_generate_v1(),
+    username varchar(512) NOT NULL UNIQUE,
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE IF NOT EXISTS notification_types (
+    id UUID NOT NULL DEFAULT uuid_generate_v1(),
+    name varchar(32) NOT NULL UNIQUE,
+    PRIMARY KEY (id)
+);
+
+CREATE TABLE IF NOT EXISTS notifications (
+    id UUID DEFAULT uuid_generate_v1(),
+    notification_type_id uuid NOT NULL REFERENCES notification_types(id),
+    user_id uuid NOT NULL REFERENCES users(id) ON DELETE CASCADE,
+    subject text NOT NULL,
+    seen boolean NOT NULL DEFAULT false,
+    deleted boolean NOT NULL DEFAULT false,
+    time_created timestamp with time zone NOT NULL DEFAULT now(),
+    message jsonb NOT NULL,
```

Comment:
I don't know if there's a specific reason you've chosen `jsonb` over `json` here. Insertion is slightly slower for `jsonb`, but it's more efficient for queries using e.g. `->` and `->>`, and it's the only one that supports indexing fields _within_ the JSON at all. But, most of the time in our code thus far, we haven't been doing json stuff in the database, preferring to parse it and work with it in our code for various reasons, which actually means the `json` type is slightly better for our use case.

Reply:
I was going with `jsonb` primarily because the documentation recommended using it over `json`. I hadn't gotten around to actually testing it with Go. I might develop a stronger opinion after some experimentation; I haven't tested `jsonb` (or heard of anyone else testing it) before now, so I hadn't developed an opinion about it yet.

I can imagine a good reason for wanting to parse `json` in our code for Clojure. For example, being able to tell `cheshire` that we want to use keywords for the map keys before doing the parsing is helpful. I'm not aware of other reasons why parsing it in our code would be preferred, however. What are some of the reasons that you're aware of?

Reply:
Well, I suppose the primary reasons are flexibility and using a more object-y vs. relation-y model. To have the DB do _all_ the parsing work, you'd need to know in advance exactly what pieces you need out of the JSON and where they live within it, and have the DB return them as columns. Perhaps the best example of this I know uses hstore (which is in ways similar) but is sort of the same idea: for OpenStreetMap map rendering, it's pulling from a big database that has lots of tags that aren't that important to rendering (for example, most maps aren't going to print out the opening hours directly on the map) but plenty that are (like what sort of object it is), so it's easiest to put all the data into a big structure in the DB and do the work in the queries. Often this sort of a context is where I see `jsonb` getting praised heavily -- basically, when you're using postgres as a JSON document store (it can often be a better one than purpose-built ones, amusingly enough, from what I hear).

It's also not too hard to imagine having the DB do part of it, and our code subsections of it -- say, if each message includes a `sent_timestamp` and `data` at the top level, or something, we could imagine doing `SELECT ..., message->'data' AS message_data, message->>'sent_timestamp'::timestamp with time zone AS sent_timestamp FROM ...`, keeping it in the `message` field to avoid dissecting it before inserting into the DB. Relatedly, this also relates to one of the cases where `jsonb` is _required_, which is if you want to index inside the JSON, so in this hypothetical, it'd have to be `jsonb` if we wanted to add an index on `message->'sent_timestamp'` to use for sorting or lookup or whatever.

But I suppose the biggest thing is just if you want to be able to fetch the JSON and then make decisions based on its structure, that rapidly becomes a huge pain to write in SQL. If most of what you're doing is more in this vein, then `jsonb` or `json` will be rather interchangeable at selection time, since either one's just going to spit out json for our code to parse.

I realize that I did neglect to mention one additional small benefit of `jsonb`, which is that it stores things somewhat smaller. Most likely, I'd imagine, not significantly enough to matter for our needs. Which may also be true of the somewhat higher performance on insert for the `json` type. And, one other almost-certainly-irrelevant bit about the types, which is if the _exact string_ you inserted is important (e.g. for matching against a hash), `json` is needed, because `jsonb` doesn't store the actual text directly.

Put more succinctly: most of the time the difference is marginal between the two (big exceptions are need for exact string matches and adding indexes for data inside the JSON structure), and elsewhere in our DBs we've used `json` because of the marginally better insert performance. I'm not opposed to `jsonb` by any means, but I like to ask about this sort of thing since I'm not fond of making these sorts of choices "by default", if that makes sense.

Reply:
Thanks for the clarification. That description fairly closely matches the impression that I got from the documentation.

To summarize, the primary benefit of `json` that I found in the documentation is that it has a slightly better insertion time compared to `jsonb`. None of our code requires whitespace preservation, duplicate keys in JSON objects, or key-order preservation, so those differences between `json` and `jsonb` are irrelevant.

The benefits of `jsonb` that were mentioned in the documentation are that it enables us to create indices on JSON fields and it improves the performance of operations on JSON documents. I also expect the slight reduction in storage usage for `jsonb` to be negligible. I don't anticipate us needing to make full use of the features of `jsonb` immediately, but it also seems like the cost for making these features available is negligible. To me, it appears that `jsonb` gives us more flexibility at very little cost.

To be honest, we're likely to extract the full JSON document and parse it in the code. I don't expect that to change anytime soon, if ever. The one potential benefit that I could see with `jsonb` that would apply to us is a slightly better performance if we decide to search for things in the JSON structure. I could envision having a `WHERE` clause that searches for specific items within the structure. Of course, that point can be countered by saying we'd probably end up doing complex searches using something like ElasticSearch instead.

As you mentioned, the two types are largely equivalent for our purposes. To me, the increased flexibility of `jsonb` appears to come at a low cost and the documentation gives a `jsonb` a modest endorsement. That's why I chose `jsonb`.